gossip4/client2/roundsubscriber.go

```diff
@@ -0,0 +1,237 @@
+	}
+
+	go func() {
+		for {
+			msg, err := sub.Next(ctx)
+			if err != nil {
+				rs.logger.Warningf("error getting sub message: %v", err)
+				return
+			}
+			if err := rs.handleMessage(ctx, msg); err != nil {
+				rs.logger.Warningf("error handling pubsub message: %v", err)
+			}
+		}
+	}()
+	return nil
+}
+
+func (rs *roundSubscriber) subscribe(subscriptionCID cid.Cid, ch chan error) subscription {
+	rs.Lock()
+	defer rs.Unlock()
+	rs.subscriptionCounter++
+	return rs.stream.Subscribe(func(evt interface{}) {
+		if evt.(cid.Cid).Equals(subscriptionCID) {
+			ch <- nil
+		}
+	})
+}
+
+func (rs *roundSubscriber) unsubscribe(sub subscription) {
+	rs.Lock()
+	defer rs.Unlock()
+	rs.subscriptionCounter--
+	rs.stream.Unsubscribe(sub)
+}
+
+func (rs *roundSubscriber) pubsubMessageToRoundConfirmation(ctx context.Context, msg *pubsub.Message) (*types.RoundConfirmation, error) {
+	bits := msg.GetData()
+	confirmation := &types.RoundConfirmation{}
+	err := cbornode.DecodeInto(bits, confirmation)
+	return confirmation, err
+}
+
+// TODO: we can cache this
+func (rs *roundSubscriber) verKeys() []*bls.VerKey {
+	keys := make([]*bls.VerKey, len(rs.group.AllSigners()))
+	for i, s := range rs.group.AllSigners() {
+		keys[i] = s.VerKey
+	}
+	return keys
+}
+
+func (rs *roundSubscriber) handleMessage(ctx context.Context, msg *pubsub.Message) error {
+	rs.logger.Debugf("handling message")
+
+	confirmation, err := rs.pubsubMessageToRoundConfirmation(ctx, msg)
+	if err != nil {
+		return fmt.Errorf("error unmarshaling: %w", err)
+	}
+
+	if rs.current != nil && confirmation.Height <= rs.current.Height {
+		return fmt.Errorf("confirmation of height %d is less than current %d", confirmation.Height, rs.current.Height)
+	}
+
+	err = sigfuncs.RestoreBLSPublicKey(confirmation.Signature, rs.verKeys())
+	if err != nil {
+		return fmt.Errorf("error retoring BLS key: %w", err)
+	}
+
+	verified, err := sigfuncs.Valid(confirmation.Signature, confirmation.CompletedRound.Bytes(), nil)
+	if !verified || err != nil {
+		return fmt.Errorf("signature invalid with error: %v", err)
+	}
+
+	rs.Lock()
+	defer rs.Unlock()
+
+	conflictSet, ok := rs.inflight[confirmation.Height]
+	if !ok {
+		conflictSet = make(roundConflictSet)
+	}
+	rs.logger.Debugf("checking quorum: %v", confirmation)
+
+	madeQuorum, updated, err := conflictSet.add(rs.group, confirmation)
+	rs.inflight[confirmation.Height] = conflictSet
+	if madeQuorum {
+		return rs.handleQuorum(ctx, updated)
+	}
+
+	return nil
+}
+
+func (rs *roundSubscriber) handleQuorum(ctx context.Context, confirmation *types.RoundConfirmation) error {
+	rs.logger.Debugf("hande Quorum: %v", confirmation)
+
+	// handleQuorum expects that it's already in a lock on the roundSubscriber
+	rs.current = confirmation
+	for key := range rs.inflight {
+		if key <= confirmation.Height {
+			delete(rs.inflight, key)
+		}
+	}
+
+	// only go and fetch the actual round and txs if we have subscribers
+	if rs.subscriptionCounter > 0 {
```

Comment:
hmmmmm. i think we can run into some issues here if we call `unsubscribe` with the same subscription more than once because we're blindly decrementing the counter with each call to `unsubscribe` above. i'm not sure how to get around that because every field on both the `Subscription` and `EventStream` types that could help are private, and there also aren't any accessor methods that i could find that would help either. 

unless i'm missing something, i think that perhaps we should take the performance hit of fetching the rounds every time instead of risking nasty bugs later.

Reply:
yup - completely true!

Reply:
oh and also pulling down the completed rounds is helpful for network resiliency and cacheing for bitswap too so a good win.